### ISSUE TITLE
`status` segment behavior with [[

### ISSUE THREAD

**Issue by jshort** [bhilburn/powerlevel9k#749]
`[[` doesn't fire the `status` segment:

```
 ❯ [[ 1 == 1 ]]                                                                                                        11:10:11  02.14.18  48% (2:09) ▅
 ✔  me@host  ~ 
 ❯ [[ 1 == 2 ]]                                                                                                        11:10:15  02.14.18  48% (2:09) ▅
 ✔  me@host  ~ 
 ❯ [[ 1 == 1 ]] ; echo $?                                                                                              11:10:18  02.14.18  48% (2:09) ▅
0
 ✔  me@host  ~ 
 ❯ [[ 1 == 2 ]] ; echo $?                                                                                              11:10:27  02.14.18  48% (2:09) ▅
1
 ✔  me@host  ~ 
 ❯
```

Also if you force a non-zero return with something else (`[` in this case) then future calls to true (zero  exit) calls of `[[` retain the previous status return code:

```
 ✔  me@host  ~ 
 ❯ [ 1 = 2 ]                                                                                                           11:14:13  02.14.18  47% (2:39) ▅
 ↵ 1  me@host  ~ 
 ❯ [[ 1 == 1 ]]                                                                                                        11:14:18  02.14.18  47% (2:39) ▅
 ↵ 1  me@host  ~ 
 ❯ [[ 1 == 1 ]]                                                                                                        11:14:25  02.14.18  47% (2:39) ▅
 ↵ 1  me@host  ~ 
```

The odd thing is that `echo $?` returns the correct exit code when you use `[[` but the status segment doesn't seem to see it.

**Comment by weilbith** [comment]
I think u r not aware about what the status segment does.
The [implementation](https://github.com/bhilburn/powerlevel9k/blob/master/powerlevel9k.zsh-theme) in function `prompt_status()` is using `RETVAL`. It can be read as `RETurn VALue`. All of ur snippets are successful and so their return value is `0` by default. If u would write a short script and add a single line `exit 1`, make it executable and run it, u should get the `FAIL_ICON`.

Quick experiment (that show fail status):
```
echo "exit 1" >> test.sh
chmod +x test.sh
./test.sh
rm test.sh
```

**Comment by jshort** [comment]
Please read my post:

```
 ❯ [[ 1 == 2 ]]                                                                                                        11:10:15  02.14.18  48% (2:09) ▅
 ✔  me@host  ~ 
1
```
Note that is a false test and therefore [[ should return non zero.  However, as you can see the check mark indicates that the status segment saw an exit code of 0.

Further evidence:

```
 ❯ [[ 1 == 2 ]] ; echo $?                                                                                              11:10:27  02.14.18  48% (2:09) ▅
1
```

So in this experiment the exit code is, in fact, 1 for the test despite the fact that the status segment is reporting 0.

**Comment by bhilburn** [comment]
@jshort - Yup, you're right. Check it out:

<img width="356" alt="screen shot 2018-03-25 at 21 03 00" src="https://user-images.githubusercontent.com/569571/37882303-3b4c02fa-3070-11e8-8e99-8682809e0900.png">

Our fancy return code / error code stuff was contributed by @docwhat in [this commit](https://github.com/bhilburn/powerlevel9k/commit/543bd2f844c561dea4e7900dd934a007b737331f#diff-ea5a405f0331722fc10a832209ffe20bR1029). I tried looking for more information on the `RETVALS` array, but couldn't seem to find anything useful. Pinging @docwhat to get some guidance =)

**Comment by docwhat** [comment]
`RETVALS` is just `"$pipestatus[@]"`

> `pipestatus <S> <Z>`
>
> An array containing the exit statuses returned by all commands in the last pipeline.

And also related:

> `PIPE_FAIL`
> By default, when a pipeline exits the exit status recorded by the shell and returned by the shell variable `$?` reflects that of the rightmost  element  of  a  pipeline.   If  this  option  is  set,  the exit status instead reflects the status of the rightmost element of the pipeline that was non-zero, or zero if all elements exited with zero status.

As @jshort mentioned, this behaves weirdly:

```
$ [[ a == b ]] ; print -l "\$? $?" "$pipestatus[@]"
$? 1
0
```

I _think_ its actually a bug of some sort in ZSH itself.  We might be able to partially work around it... something like if length of `RETVALS` is 1 and `RETVALS[1]` is 0 then use `RETVAL` instead.

**Comment by docwhat** [comment]
I've tested it in zsh 5.3 and 5.4.2 and both are the same.

**Comment by bhilburn** [comment]
@docwhat - If length of `RETVALS` is `1`, why not just always use `RETVAL`? That will simplify the conditional a bit, but still allow us to use the piped exit conditions. Thoughts?

**Comment by docwhat** [comment]
That would be a potential work-around.

**Comment by bhilburn** [comment]
Fixed by @docwhat's PR in #789!
